graders/c/Dockerfile

```diff
@@ -1,16 +1,17 @@
-FROM ubuntu:22.04
+FROM ubuntu:24.04
 
 LABEL maintainer="jonatan@yorku.ca"
 
 ARG DEBIAN_FRONTEND=noninteractive
 
-RUN apt-get update
-RUN apt-get install -y python3.10 gcc make python3-pip valgrind check pkg-config && apt clean
+RUN apt-get update && \
+    apt-get install -y python3.12 gcc make python3-pip valgrind check pkg-config && \
+    apt-get clean
 
-RUN ln -sf /usr/bin/python3.10 /usr/bin/python3
+RUN ln -sf /usr/bin/python3.12 /usr/bin/python3
 
 COPY requirements.txt /requirements.txt
-RUN pip3 install --no-cache-dir -r /requirements.txt
+RUN pip3 install --no-cache-dir --break-system-packages -r /requirements.txt
```

Comment:
Disregard if this would cause backwards incompatibility, but I think you can avoid having to add this flag (and possibly other internal issues) by creating a virtual environment and executing this Python code from there: https://docs.python.org/3/library/venv.html

Reply:
One concern I have is that the example course has test.py files with a shebang pointing to `/usr/bin/python3`, so in order to get this to work you'd need to make sure that `/usr/bin/python3` points to the venv version of python, which if I understand it correctly kind of defeats the purpose. In most cases this will not actually take effect because, by default, the  entrypoint is typically `python3 /grade/tests/test.py`, so you could theoretically just make sure that the venv version is first in the PATH, but since this is a controlled environment where you're not expected to run anything else other than the autograder, I don't see a reason to go through the trouble, unless you know something I don't about it.